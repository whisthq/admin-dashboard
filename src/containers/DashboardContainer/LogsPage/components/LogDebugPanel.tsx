import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import {
    faCircleNotch,
    faExclamation,
    faCheck,
} from '@fortawesome/free-solid-svg-icons'
import Popup from 'reactjs-popup'

import 'react-tabs/style/react-tabs.css'

import '../../../../static/App.css'

export class LogDebugPanel extends React.Component<any, any> {
    render() {
        // these exist for readability
        let filename = this.props.filename
        let log_analysis = this.props.log_analysis
        let sender = this.props.sender
        let username = this.props.username
        let connection_id = this.props.connection_id

        let query_by = username.concat('_', connection_id)

        if (
            !filename ||
            (log_analysis &&
                log_analysis[query_by] &&
                log_analysis[query_by][sender] &&
                log_analysis[query_by][sender].debug)
        ) {
            if (!filename) {
                return (
                    <div
                        style={{
                            marginLeft: 10,
                            display: 'inline-block',
                            padding: '8px 20px',
                            color: '#555555',
                            background: 'rgba(132, 132, 138, 0.1)',
                            fontSize: 13,
                            borderRadius: 3,
                            marginBottom: 10,
                        }}
                    >
                        <FontAwesomeIcon
                            icon={faExclamation}
                            style={{
                                marginRight: 10,
                                fontSize: 12,
                            }}
                        />
                        No {sender} logs
                    </div>
                )
            } else if (
                !log_analysis[query_by][sender].debug ||
                log_analysis[query_by][sender].debug.number_of_errors === 0
            ) {
                return (
                    <div
                        style={{
                            marginLeft: 10,
                            display: 'inline-block',
                            padding: '8px 20px',
                            color: '#31de4b',
                            background: 'rgba(49, 222, 75, 0.1)',
                            fontSize: 13,
                            borderRadius: 3,
                            marginBottom: 10,
                        }}
                    >
                        <FontAwesomeIcon
                            icon={faCheck}
                            style={{
                                marginRight: 10,
                                fontSize: 12,
                            }}
                        />
                        No {sender} errors
                    </div>
                )
            } else {
                return (
                    <Popup
                        trigger={
                            <div
                                style={{
                                    marginLeft: 10,
                                    display: 'inline-block',
                                    padding: '8px 20px',
                                    color: '#f5792c',
                                    background: 'rgba(245, 121, 44, 0.1)',
                                    fontSize: 13,
                                    borderRadius: 3,
                                    marginBottom: 10,
                                }}
                                className="pointerOnHover"
                            >
                                <FontAwesomeIcon
                                    icon={faExclamation}
                                    style={{
                                        marginRight: 10,
                                        fontSize: 12,
                                    }}
                                />
                                {
                                    log_analysis[
                                        username.concat('_', connection_id)
                                    ][sender].debug.number_of_errors
                                }{' '}
                                {sender} error(s) (
                                {(
                                    log_analysis[
                                        username.concat('_', connection_id)
                                    ][sender].debug.error_rate * 100
                                )
                                    .toFixed(2)
                                    .toString()}{' '}
                                % error rate)
                            </div>
                        }
                        modal
                        contentStyle={{
                            width: 600,
                            borderRadius: 5,
                            backgroundColor: 'white',
                            border: 'none',
                            height: 500,
                            padding: 30,
                            textAlign: 'left',
                            overflowY: 'scroll',
                            overflowX: 'scroll',
                        }}
                    >
                        <div>
                            <div
                                style={{
                                    fontWeight: 'bold',
                                    marginBottom: 15,
                                    fontSize: 18,
                                }}
                            >
                                {sender.charAt(0).toUpperCase() +
                                    sender.slice(1)}{' '}
                                Errors
                            </div>
                            {log_analysis[username.concat('_', connection_id)][
                                sender
                            ].debug.errors.map(function (
                                value: any,
                                _index: any
                            ) {
                                return <div>{value}</div>
                            })}
                        </div>
                    </Popup>
                )
            }
        } else {
            return (
                <div
                    style={{
                        paddingLeft: 10,
                    }}
                >
                    <div
                        style={{
                            fontWeight: 'bold',
                            marginBottom: 10,
                            marginTop: 10,
                        }}
                    >
                        {this.props.title}
                    </div>
                    <div style={{ fontSize: 13 }}>
                        <FontAwesomeIcon
                            icon={faCircleNotch}
                            spin
                            style={{
                                marginRight: 10,
                                fontSize: 12,
                                color: '#4b3ba8',
                            }}
                        />
                        Scanning {sender} logs for errors
                    </div>
                </div>
            )
        }
    }
}

export default LogDebugPanel
