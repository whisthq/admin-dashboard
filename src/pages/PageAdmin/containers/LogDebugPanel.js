import React, { Component } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import {
    faCircleNotch,
    faExclamation,
    faCheck,
} from '@fortawesome/free-solid-svg-icons'
import Popup from 'reactjs-popup'

import 'react-tabs/style/react-tabs.css'

import 'static/App.css'

class LogDebugPanel extends Component {
    render() {
        if (
            !this.props.filename ||
            (this.props.log_analysis &&
                this.props.log_analysis[
                    this.props.username.concat('_', this.props.connection_id)
                ] &&
                this.props.log_analysis[
                    this.props.username.concat('_', this.props.connection_id)
                ][this.props.sender] &&
                this.props.log_analysis[
                    this.props.username.concat('_', this.props.connection_id)
                ][this.props.sender].debug)
        ) {
            if (!this.props.filename) {
                return (
                    <div
                        style={{
                            marginLeft: 10,
                            display: 'inline-block',
                            padding: '8px 20px',
                            color: '#555555',
                            background: 'rgba(132, 132, 138, 0.1)',
                            fontSize: 13,
                            borderRadius: 3,
                            marginBottom: 10,
                        }}
                    >
                        <FontAwesomeIcon
                            icon={faExclamation}
                            style={{
                                marginRight: 10,
                                fontSize: 12,
                            }}
                        />
                        No {this.props.sender} logs
                    </div>
                )
            } else if (
                !this.props.log_analysis[
                    this.props.username.concat('_', this.props.connection_id)
                ][this.props.sender].debug ||
                this.props.log_analysis[
                    this.props.username.concat('_', this.props.connection_id)
                ][this.props.sender].debug.number_of_errors === 0
            ) {
                return (
                    <div
                        style={{
                            marginLeft: 10,
                            display: 'inline-block',
                            padding: '8px 20px',
                            color: '#31de4b',
                            background: 'rgba(49, 222, 75, 0.1)',
                            fontSize: 13,
                            borderRadius: 3,
                            marginBottom: 10,
                        }}
                    >
                        <FontAwesomeIcon
                            icon={faCheck}
                            style={{
                                marginRight: 10,
                                fontSize: 12,
                            }}
                        />
                        No {this.props.sender} errors
                    </div>
                )
            } else {
                return (
                    <Popup
                        trigger={
                            <div
                                style={{
                                    marginLeft: 10,
                                    display: 'inline-block',
                                    padding: '8px 20px',
                                    color: '#f5792c',
                                    background: 'rgba(245, 121, 44, 0.1)',
                                    fontSize: 13,
                                    borderRadius: 3,
                                    marginBottom: 10,
                                }}
                                className="pointerOnHover"
                            >
                                <FontAwesomeIcon
                                    icon={faExclamation}
                                    style={{
                                        marginRight: 10,
                                        fontSize: 12,
                                    }}
                                />
                                {
                                    this.props.log_analysis[
                                        this.props.username.concat(
                                            '_',
                                            this.props.connection_id
                                        )
                                    ][this.props.sender].debug.number_of_errors
                                }{' '}
                                {this.props.sender} error(s) (
                                {(
                                    this.props.log_analysis[
                                        this.props.username.concat(
                                            '_',
                                            this.props.connection_id
                                        )
                                    ][this.props.sender].debug.error_rate * 100
                                )
                                    .toFixed(2)
                                    .toString()}{' '}
                                % error rate)
                            </div>
                        }
                        modal
                        contentStyle={{
                            width: 600,
                            borderRadius: 5,
                            backgroundColor: 'white',
                            border: 'none',
                            height: 500,
                            padding: 30,
                            textAlign: 'left',
                            overflowY: 'scroll',
                            overflowX: 'scroll',
                        }}
                    >
                        <div>
                            <div
                                style={{
                                    fontWeight: 'bold',
                                    marginBottom: 15,
                                    fontSize: 18,
                                }}
                            >
                                {this.props.sender.charAt(0).toUpperCase() +
                                    this.props.sender.slice(1)}{' '}
                                Errors
                            </div>
                            {this.props.log_analysis[
                                this.props.username.concat(
                                    '_',
                                    this.props.connection_id
                                )
                            ][this.props.sender].debug.errors.map(function (
                                value,
                                index
                            ) {
                                return <div>{value}</div>
                            })}
                        </div>
                    </Popup>
                )
            }
        } else {
            return (
                <div
                    style={{
                        paddingLeft: 10,
                    }}
                >
                    <div
                        style={{
                            fontWeight: 'bold',
                            marginBottom: 10,
                            marginTop: 10,
                        }}
                    >
                        {this.props.title}
                    </div>
                    <div style={{ fontSize: 13 }}>
                        <FontAwesomeIcon
                            icon={faCircleNotch}
                            spin
                            style={{
                                marginRight: 10,
                                fontSize: 12,
                                color: '#4b3ba8',
                            }}
                        />
                        Scanning {this.props.sender} logs for errors
                    </div>
                </div>
            )
        }
    }
}

export default LogDebugPanel
