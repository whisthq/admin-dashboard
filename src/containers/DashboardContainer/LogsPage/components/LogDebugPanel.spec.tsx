import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import toJson from 'enzyme-to-json'

import { LogDebugPanel } from './LogDebugPanel'

Enzyme.configure({ adapter: new Adapter() })

/*
The Log Debug panel is basically a small icon + text box in the log that tells us how many errors happened.
It depends purely on props. If it gets props and they have a corresponding filename it returns either a green
check with a message for no errors, or a red exclamation point with the error rate. If there is nothing (no logs)
for that sender (client or server) it displays an icon and helpful message.

If the props are null it displays a loading icon and says that it is loading.
 */
describe('<LogDebugPanel />', () => {
    // snapshot test
    it('renders correctly', () => {
        const wrapper = shallow(
            <LogDebugPanel
                // buncha dummy data
                dispatch={(items: any) => {
                    // mock
                }}
                title="title"
                filename="filename"
                username="user"
                sender="sender"
                connection_id="id"
                metric="metric"
                logAnalysis={{
                    user_id: {
                        sender: {
                            metric: {
                                output: [1], // not enough data to plot
                            },
                        },
                    },
                }}
            />
        )

        expect(toJson(wrapper)).toMatchSnapshot()
    })

    it('displays an icon (always)', () => {
        const wrapper = shallow(
            <LogDebugPanel
                // buncha dummy data
                dispatch={(items: any) => {
                    // mock
                }}
                title="title"
                filename="filename"
                username="user"
                sender="sender"
                connection_id="id"
                metric="metric"
                logAnalysis={{
                    user_id: {
                        sender: {
                            metric: {
                                output: [1], // not enough data to plot
                            },
                        },
                    },
                }}
            />
        )

        expect(wrapper.find(FontAwesomeIcon)).toHaveLength(1)
    })

    // TODO (optional) test for string matching messages
    // since we don't want to pin down the message in case we want to change it for now, we are keeping
    // from implementing it it this way
})
