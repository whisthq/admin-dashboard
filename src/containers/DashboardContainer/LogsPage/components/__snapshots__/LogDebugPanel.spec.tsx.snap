// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LogDebugPanel /> renders correctly 1`] = `
<div
  style={
    Object {
      "paddingLeft": 10,
    }
  }
>
  <div
    style={
      Object {
        "fontWeight": "bold",
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  >
    title
  </div>
  <div
    style={
      Object {
        "fontSize": 13,
      }
    }
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f1ce",
            "M288 39.056v16.659c0 10.804 7.281 20.159 17.686 23.066C383.204 100.434 440 171.518 440 256c0 101.689-82.295 184-184 184-101.689 0-184-82.295-184-184 0-84.47 56.786-155.564 134.312-177.219C216.719 75.874 224 66.517 224 55.712V39.064c0-15.709-14.834-27.153-30.046-23.234C86.603 43.482 7.394 141.206 8.003 257.332c.72 137.052 111.477 246.956 248.531 246.667C393.255 503.711 504 392.788 504 256c0-115.633-79.14-212.779-186.211-240.236C302.678 11.889 288 23.456 288 39.056z",
          ],
          "iconName": "circle-notch",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={true}
      style={
        Object {
          "color": "#4b3ba8",
          "fontSize": 12,
          "marginRight": 10,
        }
      }
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
    Scanning 
    sender
     logs for errors
  </div>
</div>
`;
